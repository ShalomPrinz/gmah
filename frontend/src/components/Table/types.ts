interface TableColumn {
  id: number;
  label?: string;
  path: string;
}

export type { TableColumn };
