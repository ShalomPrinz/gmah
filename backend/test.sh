#!/bin/bash

if [[ $1 == "--coverage" ]]; then
    coverage run -m unittest discover tests
    coverage html
else
    python -m unittest discover tests
fi
