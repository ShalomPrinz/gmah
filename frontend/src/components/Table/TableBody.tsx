import { ComponentType } from "react";
import { ConditionalList } from "../";

import type { TableColumn } from "./types";

interface TableData {
  [key: string]: string;
}

/** Each column in this dictionary is rendered with its callback result */
type ColumnMapper = {
  [column: string]: (value: any) => {
    background?: string | undefined;
    text?: string | undefined;
  };
};

export interface TableBodyProps {
  columnMapper?: ColumnMapper;
  columns: TableColumn[];
  data: TableData[];
  dataIdProp: string;
  LastColumn?: ComponentType<{ item: TableData }>;
  numberedTable?: boolean;
}

const TableBody = ({
  columnMapper = {},
  columns,
  data,
  dataIdProp,
  LastColumn,
  numberedTable,
}: TableBodyProps) => {
  const columnCallback = (item: TableData, column: TableColumn) => {
    const baseClassName = "align-middle";

    if (column.path in columnMapper) {
      const { background, text } = columnMapper[column.path](item[column.path]);
      if (typeof background === "string")
        return <td className={`${baseClassName} bg-${background}`}></td>;
      if (typeof text === "string")
        return <td className={baseClassName}>{text}</td>;
    }

    return (
      <td className={baseClassName}>
        {Object.hasOwn(item, column.path) ? item[column.path] : "שגיאה"}
      </td>
    );
  };

  const rowCallback = (item: TableData, rowIndex: number) => (
    <tr key={item[dataIdProp]}>
      {numberedTable && (
        <td className="pe-4 align-middle fw-bold">{rowIndex + 1}</td>
      )}
      <ConditionalList
        itemCallback={(column: TableColumn) => columnCallback(item, column)}
        list={columns}
      />
      {LastColumn && <td>{<LastColumn item={item} />}</td>}
    </tr>
  );

  return <tbody>{data.map((item, index) => rowCallback(item, index))}</tbody>;
};

export default TableBody;
