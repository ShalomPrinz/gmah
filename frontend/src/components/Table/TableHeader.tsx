import { ConditionalList } from "../";
import type { TableColumn } from "./types";

const headerCallback = ({ label }: TableColumn) => (
  <th className="p-3">{label}</th>
);

interface TableHeaderProps {
  columns: TableColumn[];
}

const TableHeader = ({ columns }: TableHeaderProps) => {
  return (
    <thead>
      <tr>
        <ConditionalList itemCallback={headerCallback} list={columns} />
      </tr>
    </thead>
  );
};

export default TableHeader;
