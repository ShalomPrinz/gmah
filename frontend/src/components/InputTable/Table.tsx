import { FieldArray, Form, Formik } from "formik";
import type { FormikErrors } from "formik";
import { useEffect, useState } from "react";
import { toast } from "react-toastify";

import TableBody from "./TableBody";
import TableHeader from "./TableHeader";

import type { TableColumn } from "./types";
import IconComponent from "../Icon";

type TableRow = {
  [key: string]: string;
};

interface InputTableProps {
  columns: TableColumn[];
  /** Used for adding item to the table */
  emptyItem: TableRow;
  initialData: TableRow[];
  inputsName: string;
  onSubmit: (data: TableRow[]) => Promise<boolean>;
  /** Yup Object Schema with one key value pair: { [inputsName]: data } */
  schema: any;
}

const InputTable = ({
  columns,
  emptyItem,
  initialData,
  inputsName,
  onSubmit,
  schema,
}: InputTableProps) => {
  const { itemsCount, addItem, removeItem } = useItemsCount(
    initialData.length,
    emptyItem
  );
  const displayTable = itemsCount > 0;

  const toastUnknownSubmitError = () =>
    toast.error("קרתה תקלה בהוספת המשפחות", { toastId: "addFamiliesError" });

  const onSubmitResolved = (isSuccess: boolean) => {
    if (isSuccess) {
      toast.success(`הוספת בהצלחה ${itemsCount} משפחות חדשות לגמ"ח`);
    } else {
      toastUnknownSubmitError();
    }
  };

  function handleErrorSubmission(
    errors: FormikErrors<{ [x: string]: TableRow[] }>
  ) {
    if (!errors || !errors[inputsName] || !Array.isArray(errors[inputsName]))
      return;

    const errorsArray = errors[inputsName] as TableRow[];
    errorsArray.forEach((familyErrors, index) => {
      familyErrors &&
        Object.entries(familyErrors).forEach(([field, message]) =>
          toast.warning(
            `יש שגיאה ב${field} של משפחה מספר ${index + 1}: ${message}`,
            { toastId: `${index}:${field}` }
          )
        );
    });
  }

  return (
    <Formik
      enableReinitialize
      initialValues={{ [inputsName]: initialData }}
      validationSchema={schema}
      onSubmit={(values, { setSubmitting }) => {
        setSubmitting(true);
        onSubmit(values[inputsName])
          .then((isSuccess) => onSubmitResolved(isSuccess))
          .catch(() => toastUnknownSubmitError())
          .finally(() => setSubmitting(false));
      }}
    >
      {({ isSubmitting, errors }) => (
        <Form onSubmitCapture={() => handleErrorSubmission(errors)}>
          <FieldArray name={inputsName}>
            {({ push, remove }) => (
              <>
                <table className="bg-white">
                  <TableHeader columns={columns} display={displayTable} />
                  <TableBody
                    columns={columns}
                    inputsName={inputsName}
                    items={itemsCount}
                    removeFunc={(index: number) => removeItem(remove, index)}
                  />
                </table>
                <button
                  className="fs-3 bg-default rounded p-4 ms-5 mb-5 position-fixed bottom-0 start-0 button-hover"
                  onClick={() => addItem(push)}
                  type="button"
                >
                  <span className="ps-3">הוסף שורה בטבלה</span>
                  <IconComponent icon="addFamily" />
                </button>
              </>
            )}
          </FieldArray>
          {isSubmitting && (
            <span className="fs-4">הוספת המשפחות מתבצעת כעת, אנא המתן...</span>
          )}
          {displayTable && (
            <button
              className="bg-default p-4 rounded fs-3 button-border-focus m-5"
              disabled={isSubmitting}
              type="submit"
            >
              הוסף משפחות
            </button>
          )}
          {!displayTable && (
            <h1 className="mt-5 pt-4 fw-light">
              - כאן תוכל לערוך את המשפחות לאחר שתדביק את הטבלה -
            </h1>
          )}
        </Form>
      )}
    </Formik>
  );
};

function useItemsCount(initialCount: number, emptyItem: TableRow) {
  const [itemsCount, setItemsCount] = useState(initialCount);

  useEffect(() => setItemsCount(initialCount), [initialCount]);

  const addItem = (pushFunc: (obj: any) => void) => {
    pushFunc(emptyItem);
    setItemsCount((count) => count + 1);
  };

  const removeItem = (removeFunc: (index: number) => void, index: number) => {
    removeFunc(index);
    setItemsCount((count) => count - 1);
  };

  return { itemsCount, addItem, removeItem };
}

export default InputTable;
