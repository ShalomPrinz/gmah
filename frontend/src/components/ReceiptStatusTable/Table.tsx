import { useEffect } from "react";
import { FormProvider, useFieldArray, useForm } from "react-hook-form";

import TableBody from "./TableBody";
import TableHeader from "./TableHeader";
import type { FormItem, FormValues } from "./types";
import { prepareKey } from "../../util";

interface ReceiptStatusTableProps {
  initialValues: FormItem[];
  formName: string;
  onSubmit: (values: FormItem[]) => void;
}

function ReceiptStatusTable({
  formName,
  initialValues,
  onSubmit,
}: ReceiptStatusTableProps) {
  const name = prepareKey(formName);
  const defaultValues = {
    [name]: initialValues,
  };

  const formMethods = useForm<FormValues>({
    defaultValues,
    mode: "onSubmit",
  });

  useEffect(() => formMethods.reset(defaultValues), [name, initialValues]);

  // All form methods are required for Form Context Provider
  const {
    control,
    handleSubmit,
    formState: { isSubmitting },
    getValues,
    reset,
  } = formMethods;

  const { fields } = useFieldArray({
    name: name,
    control,
  });

  function setAllStatuses(status: boolean) {
    const newFormState = getValues()[name].map((receipt) => ({
      ...receipt,
      status,
    }));
    reset({ [name]: newFormState });
  }

  function setAllDates(date: string) {
    const newFormState = getValues()[name].map((receipt) => ({
      ...receipt,
      date,
    }));
    reset({ [name]: newFormState });
  }

  function onSubmitInit(e: React.FormEvent<HTMLFormElement>) {
    e.preventDefault();
    handleSubmit((data) => onSubmit(data[name]))();
  }

  return (
    <FormProvider {...formMethods}>
      <form onSubmit={onSubmitInit}>
        <h2 className="d-flex justify-content-between">
          <span className="fs-1 fw-bold bottom-border">{formName}</span>
          <button
            className="bg-default fs-4 rounded mx-5 px-3 py-2"
            disabled={isSubmitting}
            type="submit"
          >
            שמור שינויים
          </button>
        </h2>
        <table className="table rounded bg-white">
          <TableHeader
            key={formName}
            setAllDates={setAllDates}
            setAllStatuses={setAllStatuses}
          />
          <TableBody fields={fields} formName={name} />
        </table>
      </form>
    </FormProvider>
  );
}

export default ReceiptStatusTable;
