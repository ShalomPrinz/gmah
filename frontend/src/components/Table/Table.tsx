import TableBody, { TableBodyProps } from "./TableBody";
import TableHeader, { TableHeaderProps } from "./TableHeader";

type TableProps = TableHeaderProps & TableBodyProps;

const Table = ({ headerHighlight, ...props }: TableProps) => {
  const hasLastColumn = typeof props.LastColumn !== "undefined";

  return (
    <table className="table rounded bg-white">
      <TableHeader
        columns={props.columns}
        hasLastColumn={hasLastColumn}
        headerHighlight={headerHighlight}
      />
      <TableBody {...props} />
    </table>
  );
};

export default Table;
