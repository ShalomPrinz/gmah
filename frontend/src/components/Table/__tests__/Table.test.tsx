import { describe, expect, it } from "vitest";

import { render } from "../../../../test";

import Table from "../Table";

describe("Table", () => {
  describe("snapshot", () => {
    it("should render empty table", () => {
      const { asFragment } = render(
        <Table columns={[]} data={[]} dataIdProp="" />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render table with one column", () => {
      const { asFragment } = render(
        <Table
          columns={[
            {
              id: 0,
              path: "name",
            },
          ]}
          data={[]}
          dataIdProp=""
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render table with three columns", () => {
      const { asFragment } = render(
        <Table
          columns={[
            { id: 0, path: "name" },
            { id: 1, path: "title", label: "TITLE" },
            { id: 2, path: "property" },
          ]}
          data={[]}
          dataIdProp=""
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render table with one column and one data row", () => {
      const { asFragment } = render(
        <Table
          columns={[{ id: 0, path: "name" }]}
          data={[
            {
              name: "Shalom",
            },
          ]}
          dataIdProp="name"
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render table with two labeled columns and one data row", () => {
      const { asFragment } = render(
        <Table
          columns={[
            { id: 0, label: "First Name", path: "name" },
            { id: 1, label: "Last Name", path: "family" },
          ]}
          data={[
            {
              id: "0",
              name: "Shalom",
              family: "Prinz",
            },
            {
              id: "1",
              name: "Yosi",
              family: "Cohen",
            },
          ]}
          dataIdProp="id"
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render table with two columns and three data rows", () => {
      const { asFragment } = render(
        <Table
          columns={[
            { id: 0, path: "title" },
            { id: 1, path: "description" },
          ]}
          data={[
            {
              id: "0",
              title: "Shalom",
              description: "this is my first name",
            },
            {
              id: "1",
              title: "Prinz",
              description: "this is my last name",
            },
          ]}
          dataIdProp="id"
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render table with one column, one data row, and one special column", () => {
      const { asFragment } = render(
        <Table
          columns={[{ id: 0, path: "value" }]}
          data={[
            {
              id: "0",
              value: "result",
            },
          ]}
          dataIdProp="id"
          LastColumn={({ item }) => <button>Click Me! {item.value}</button>}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render table with two columns and highlight the second", () => {
      const { asFragment } = render(
        <Table
          columns={[
            { id: 0, path: "value" },
            { id: 1, path: "content" },
          ]}
          data={[]}
          dataIdProp=""
          headerHighlight="content"
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
