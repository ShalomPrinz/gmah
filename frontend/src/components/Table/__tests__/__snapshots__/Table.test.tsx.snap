// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > snapshot > should render empty table 1`] = `
<DocumentFragment>
  <table
    class="table rounded bg-white"
  >
    <thead>
      <tr />
    </thead>
    <tbody />
  </table>
</DocumentFragment>
`;

exports[`Table > snapshot > should render table with one column 1`] = `
<DocumentFragment>
  <table
    class="table rounded bg-white"
  >
    <thead>
      <tr>
        <th
          class="fs-5 p-3"
        >
          name
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</DocumentFragment>
`;

exports[`Table > snapshot > should render table with one column and one data row 1`] = `
<DocumentFragment>
  <table
    class="table rounded bg-white"
  >
    <thead>
      <tr>
        <th
          class="fs-5 p-3"
        >
          name
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="align-middle"
        >
          Shalom
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Table > snapshot > should render table with one column, one data row, and one special column 1`] = `
<DocumentFragment>
  <table
    class="table rounded bg-white"
  >
    <thead>
      <tr>
        <th
          class="fs-5 p-3"
        >
          value
        </th>
        <th />
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="align-middle"
        >
          result
        </td>
        <td>
          <button>
            Click Me! result
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Table > snapshot > should render table with three columns 1`] = `
<DocumentFragment>
  <table
    class="table rounded bg-white"
  >
    <thead>
      <tr>
        <th
          class="fs-5 p-3"
        >
          name
        </th>
        <th
          class="fs-5 p-3"
        >
          TITLE
        </th>
        <th
          class="fs-5 p-3"
        >
          property
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</DocumentFragment>
`;

exports[`Table > snapshot > should render table with two columns and highlight the second 1`] = `
<DocumentFragment>
  <table
    class="table rounded bg-white"
  >
    <thead>
      <tr>
        <th
          class="fs-5 p-3"
        >
          value
        </th>
        <th
          class="fs-5 p-3 bg-warning"
        >
          content
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</DocumentFragment>
`;

exports[`Table > snapshot > should render table with two columns and three data rows 1`] = `
<DocumentFragment>
  <table
    class="table rounded bg-white"
  >
    <thead>
      <tr>
        <th
          class="fs-5 p-3"
        >
          title
        </th>
        <th
          class="fs-5 p-3"
        >
          description
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="align-middle"
        >
          Shalom
        </td>
        <td
          class="align-middle"
        >
          this is my first name
        </td>
      </tr>
      <tr>
        <td
          class="align-middle"
        >
          Prinz
        </td>
        <td
          class="align-middle"
        >
          this is my last name
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Table > snapshot > should render table with two labeled columns and one data row 1`] = `
<DocumentFragment>
  <table
    class="table rounded bg-white"
  >
    <thead>
      <tr>
        <th
          class="fs-5 p-3"
        >
          First Name
        </th>
        <th
          class="fs-5 p-3"
        >
          Last Name
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="align-middle"
        >
          Shalom
        </td>
        <td
          class="align-middle"
        >
          Prinz
        </td>
      </tr>
      <tr>
        <td
          class="align-middle"
        >
          Yosi
        </td>
        <td
          class="align-middle"
        >
          Cohen
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
